openstack/tools/charmed_openstack_functest_runner.sh

```diff
@@ -0,0 +1,121 @@
+
+CIDR_OAM=`openstack subnet show subnet_${OS_USERNAME}-psd -c cidr -f value`
+OAM_MAX=$(ipcalc $CIDR_OAM| awk '$1=="HostMax:" {print $2}')
+OAM_MIN=$(ipcalc $CIDR_OAM| awk '$1=="HostMin:" {print $2}')
+OAM_MIN_ABC=${OAM_MIN%.*}
+OAM_MAX_D=${OAM_MAX##*.}
+# Picking last two addresses and hoping they dont get used by Neutron.
+export OS_VIP00=${OAM_MIN_ABC}.$(($OAM_MAX_D - 1))
+export OS_VIP01=${OAM_MIN_ABC}.$(($OAM_MAX_D - 2))
+
+# More information on config https://github.com/openstack-charmers/zaza/blob/master/doc/source/runningcharmtests.rst
+export {,TEST_}NET_ID=$(openstack network show net_${OS_USERNAME}-psd-extra -f value -c id)
+export {,TEST_}FIP_RANGE=$FIP_MIN:$FIP_MAX
+export {,TEST_}GATEWAY=$(openstack subnet show subnet_${OS_USERNAME}-psd-extra -c gateway_ip -f value)
+export {,TEST_}NAME_SERVER=91.189.91.131
+export {,TEST_}CIDR_PRIV=192.168.21.0/24
+#export SWIFT_IP=10.140.56.22
+export TEST_MODEL_SETTINGS="image-stream=released;default-series=jammy;test-mode=true;transmit-vendor-metrics=false"
+# We need to set TEST_JUJU3 as well as the constraints file
+# Ref: https://github.com/openstack-charmers/zaza/blob/e96ab098f00951079fccb34bc38d4ae6ebb38606/setup.py#L47
+export TEST_JUJU3=1
+
+# NOTE: this should not be necessary for > juju 2.x but since we still have a need for it we add it in
+export TEST_ZAZA_BUG_LP1987332=1
+
+# Some charms point to an upstream constraints file that installs python-libjuju 2.x so we need to do this to ensure we get 3.x
+export TEST_CONSTRAINTS_FILE=https://raw.githubusercontent.com/openstack-charmers/zaza/master/constraints-juju34.txt
+
+# 2. Build
+if ! $SKIP_BUILD; then
+    sudo snap refresh charmcraft --channel $(grep charmcraft_channel osci.yaml| sed -r 's/.+:\s+(\S+)/\1/')
+
+    # ensure lxc initialised
+    lxd init --auto || true
+
+    tox -re build
+fi
+
+# 3. Run functional tests.
+
+# If a func test pr is provided switch to that pr.
+if [[ -n $FUNC_TEST_PR ]]; then
+    (
+    [[ -d src ]] && cd src
+    # We use the zosci-config tools to do this.
+    [[ -d ~/zosci-config ]] || ( cd; git clone https://github.com/openstack-charmers/zosci-config; )
+    (cd ~/zosci-config; git checkout master; git pull;)
+    MSG=$(echo "Func-Test-Pr: https://github.com/openstack-charmers/zaza-openstack-tests/pull/$FUNC_TEST_PR"| base64)
+    ~/zosci-config/roles/handle-func-test-pr/files/process_func_test_pr.py -f ./test-requirements.txt "$MSG"
+    )
+fi
+
+if [[ -n $FUNC_TEST_TARGET ]]; then
+  func_targets=( $FUNC_TEST_TARGET )
+else
+  func_targets=( $(python3 $(realpath $(dirname $0))/identify_charm_func_tests.py) )
+fi
+
+for target in ${func_targets[@]}; do
+    [[ -d src ]] && pushd src || true
+    tox -re func-target -- $target
+
+    read -p "Destroy model and run next test? [ENTER]"
+    # cleanup before next run
+    model=`juju list-models| egrep -o "^zaza-\S+"|tr -d '*'`
+    juju destroy-model --no-prompt $model --force --no-wait --destroy-storage
```

Comment:
This will fail if you have multiple matching models. Ideally it would also print the name of all the models it will destroy before it does, in case you had created another model with a name starting with zaza.

You can also just tell zaza to destroy the model itself automatically by passing an argument, perhaps we should use that by default and then have an option to switch to manual cleanup behaviour for trying to debug the tests?

Reply:
do you happen to know what the argument is?

Reply:
we can fix this as a future patch